# สำรวจชุดข้อมูลทิป

**จุดประสงค์การเรียนรู้**
1. การรู้จักและใช้งาน `DataFrame` และ `Series` เบื้องต้น กับชุดข้อมูลจริง
2. ตอบคำถามเกี่ยวกับสถิติการได้รับทิปจากข้อมูลที่มี
3. จัดการข้อมูลในเบื้องต้น
4. วาดกราฟเบื้องต้น ได้แก่ กราฟแท่ง, scatter plot, และ histogram ด้วย `plotly`

*ก่อนเริ่ม* นักศึกษาสามารถ google หาข้อมูลได้หรือไม่ว่า
- เว็บไซต์ทางการ (official) ของ `Pandas` คือเว็บไหน
- `Pandas` เวอร์ชันแรกถูกเผยแพร่ในปี ค.ศ. เท่าไร
- ที่มาของชื่อ `Pandas` มาจากอะไร





**เริ่มต้นบทเรียน**


พนักงานเสิร์ฟคนหนึ่งบันทึกข้อมูลเกี่ยวกับทิป ที่เขาได้รับขณะทำงานที่ร้านอาหารแห่งหนึ่งในสหรัฐอเมริกา  ซึ่งเมื่อเขาได้ทิปแต่ละครั้ง เขาจะบันทึกข้อมูลลงในไฟล์ `tips.csv` อย่างสม่ำเสมอ ตลอดช่วงเวลาหลายเดือนที่ผ่านมา โดยรายละเอียดของค่าที่เขาบันทึกแต่ละครั้ง มีดังนี้

| ค่าที่  | รายละเอียด |
|:---:|:---|
|1| บิลค่าอาหาร (ดอลลาร์) |
|2| ทิปที่เขาได้รับ (ดอลลาร์) |
|3| เพศของผู้จ่ายบิลค่าอาหาร |
|4| มีคนสูบบุหรี่หรือไม่ (ในโต๊ะของลูกค้าในบิลนั้น)  |
|5| วันที่ได้รับทิป  |
|6| เวลาที่จ่ายบิล |
|7| จำนวนลูกค้าในบิลนั้น |

พนักงานเสิร์ฟคนนี้ ได้แชร์ไฟล์ `tips.csv` มาให้นักศึกษา โดยร้องขอให้นักศึกษาศึกษาข้อมูลที่มีในเบื้องต้น เพื่อตอบคำถามที่เขาอยากรู้ ซึ่งในขั้นแรก นักศึกษาจึงได้ดาวน์โหลดและอ่านข้อมูลไฟล์ `tips.csv` เข้ามาโดยใช้คำสั่ง `read_csv()` [(คู่มือ)](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html) ของ `Pandas`ดังนี้

## **การอ่านข้อมูล**

In [ ]:
import pandas as pd
tips = pd.read_csv('https://raw.githubusercontent.com/kasemsit/269382/main/dataset/tips.csv')

In [ ]:
print(tips)

     total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
1         10.34  1.66    Male     No   Sun  Dinner     3
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       13.00  2.00  Female    Yes  Thur   Lunch     2
244       18.78  3.00  Female     No  Thur  Dinner     2

[245 rows x 7 columns]


ทั้งนี้ ไม่จำเป็นต้องใช้ฟังก์ชัน `print()` ในการพิมพ์ตารางก็ได้ โดย Google Colab สามารถแสดงผลตาราง `tips` ได้สวยงาม เพียงเรียกตัวแปร `tips` ในบรรทัดสุดท้ายของเซลล์

In [ ]:
# แสดงตาราง
tips

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
1         10.34  1.66    Male     No   Sun  Dinner     3
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       13.00  2.00  Female    Yes  Thur   Lunch     2
244       18.78  3.00  Female     No  Thur  Dinner     2

[245 rows x 7 columns]

จากข้อมูล เนื่องจากนักศึกษาไม่ได้เป็นคนเก็บข้อมูลเอง ดังนั้นเพื่อความแน่ใจ นักศึกษาจึงตรวจสอบข้อมูลที่ได้มาในเบื้องต้น ตามลำดับ ด้วยการตั้งคำถามเหล่านี้

## คำถามขั้นแรกเกี่ยวกับตัวแปรตารางที่อ่านเข้ามา


### ตัวแปร `tips` เป็นตัวแปรชนิดใดในภาษา Python

In [ ]:
x = type(tips)
print(f'ชนิดข้อมูลคือ {x}')

ชนิดข้อมูลคือ <class 'pandas.core.frame.DataFrame'>


จะเห็นว่า ตัวแปร `tips` เป็นตัวแปรชนิด `DataFrame` ที่ถูกนิยามไว้ใน `Pandas`

```{toggle}

**ทบทวน: ชนิดตัวแปรพื้นฐานในภาษา `Python`**

1) ชนิดตัวแปรที่เก็บได้ค่าเดียว

|ชนิดตัวแปร  | ความหมาย |ตัวอย่าง |
|:-- |:-- |:--
|`int` | ตัวแปรเก็บค่าจำนวนเต็ม (Interger) | `100` |
|`float` | ตัวแปรเก็บค่าจำนวนทศนิยม (Float) | `1.0` |
|`bool` | ตัวแปรเก็บค่าตรรกะ (Boolean) | `True`,`False`  |

2) ชนิดตัวแปรที่เก็บได้หลายค่า (container)

|ชนิดตัวแปร  | ความหมาย | ตัวอย่างที่มีสมาชิก 0 ตัว | ตัวอย่างที่มีสมาชิก 1 ตัว | ตัวอย่างที่มีสมาชิกหลายตัว |
|:-- |:-- |:--|:-- |:-- |
|`str` | String | '' หรือ "" | 'A' | `'123'`,`'hello'` |
|`list` | List | `[]` หรือ `list()` | `[2]` | `[3,2,'hello',`False`, 1.0]` |
|`tuple` | Tuple | `()` หรือ `tuple()` | `(2,)` | `(3,2,'hello',`False`, 1.0)`|
|`set` | Set | `set()` | `{2}` | `{3,2,'hello',`False`, 1.0}` |
| `dict` | Dictionary | `{}` หรือ `dict()` | `{2:3}` | `{1:3,2:2,'T':'hello'}` |

```



**คำถาม:** ลำดับเลข 0,1,2,...,244 ที่แสดงจากตัวแปร `tips` เป็นค่าที่เก็บในไฟล์ `csv` ใช่หรือไม่

### ข้อมูลมีกี่แถวและกี่คอลัมน์

**วิธีที่ 1** ใช้คำสั่ง `len()` นับจำนวนแถว


In [ ]:
x = len(tips)
print(f'จำนวนแถวคือ {x}')

**วิธีที่ 2** ใช้คำสั่ง `DataFrame.shape` นับจำนวนแถวและจำนวนคอลัมน์

In [ ]:
x = tips.shape
print(f'(แถว,คอลัมน์) มีจำนวนคือ {x}')

(แถว,คอลัมน์) มีจำนวนคือ (245, 7)


`tips.shape` จะให้ `tuple` ที่เก็บจำนวนแถวและจำนวนคอลัมน์ตามลำดับ

In [ ]:
n_row = tips.shape[0]
n_col = tips.shape[1]
print(f'ข้อมูลมี {n_row} แถว and {n_col} คอลัมน์')

ข้อมูลมี 245 แถว and 7 คอลัมน์


In [ ]:
n_row, n_col = tips.shape
print(f'ข้อมูลมี {n_row} แถว and {n_col} คอลัมน์')

ข้อมูลมี 245 แถว and 7 คอลัมน์



### ตัวอย่างข้อมูลบางแถวในตารางมีค่าอะไรบ้าง

**แบบที่ 1** `DataFrame.head()` เช่น ดูเฉพาะ 2 แถวแรก

In [ ]:
tips.head(2)

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3

จะเห็นว่า แถวแรกคือแถวที่ 0, แถวถัดมาคือแถวที่ 1

**แบบที่ 2** `DataFrame.tail()` เช่น ดูเฉพาะ 3 แถวสุดท้าย

In [ ]:
tips.tail(3)

total_bill   tip     sex smoker   day    time  size
242       17.82  1.75    Male     No   Sat  Dinner     2
243       13.00  2.00  Female    Yes  Thur   Lunch     2
244       18.78  3.00  Female     No  Thur  Dinner     2

จะเห็นว่า แถวสุดท้ายคือแถวที่ 243 เพราะข้อมูลเป็นตารางที่มีทั้งหมด 244 แถว นั่นเอง

**แบบที่ 3** `DataFrame.sample()` เช่น สุ่มมา 4 แถว

In [ ]:
tips.sample(4)

total_bill   tip     sex smoker  day    time  size
236       12.60  1.00    Male    Yes  Sat  Dinner     2
29        19.65  3.00  Female     No  Sat  Dinner     2
60        20.29  3.21    Male    Yes  Sat  Dinner     2
13        18.43  3.00    Male     No  Sun  Dinner     4

จะเห็นว่า ถ้าเรียกคำสั่งนี้ซ้ำ ๆ จะได้ข้อมูลที่ไม่ซ้ำแถวกัน อย่างไรก็ตาม หากต้องการให้สุ่มได้ซ้ำแบบเดิม (อาจทำเพื่อแสดงให้คนอื่นดู) ก็อาจส่งค่า `random_state` ให้ฟังก์ชัน `sample` ด้วย

In [ ]:
tips.sample(4, random_state=123)  # กำหนดค่า random_state เป็นจำนวนเต็มได้ตามต้องการ

total_bill   tip     sex smoker  day    time  size
112       38.07  4.00    Male     No  Sun  Dinner     3
19        20.65  3.35    Male     No  Sat  Dinner     3
187       30.46  2.00    Male    Yes  Sun  Dinner     5
169       10.63  2.00  Female    Yes  Sat  Dinner     2

หมายเหตุ: ฟังก์ชันของ `Pandas` (รวมถึงฟังก์ชันอื่น ๆ ที่นักศึกษาจะได้พบในอนาคต) ที่มีพฤติกรรมในลักษณะของการสุ่มค่า มักจะสามารถกำหนดค่า `random_state` เพื่อล็อกผลของการสุ่มได้ เช่น `sklearn.model_selection.train_test_split(..., random_state=8888)`  เป็นต้น

### ต้องการรายงานเกี่ยวกับคอลัมน์ในตาราง

In [ ]:
tips.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 245 entries, 0 to 244
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   total_bill  245 non-null    float64
 1   tip         245 non-null    float64
 2   sex         245 non-null    object 
 3   smoker      245 non-null    object 
 4   day         245 non-null    object 
 5   time        245 non-null    object 
 6   size        245 non-null    int64  
dtypes: float64(2), int64(1), object(4)
memory usage: 13.5+ KB


สิ่งที่คำสั่ง `DataFrame.info()` รายงานออกมา คือ
1. ชื่อคอลัมน์
2. จำนวนข้อมูลในแต่ละคอลัมน์ ไม่รวม missing value เช่นจะเห็นว่า คอลัมน์ชื่อ `total_bill` มีข้อมูลอยู่ 245 ค่า (non-null) และไม่มีข้อมูลที่สูญหาย (null) เลย
3. Dtype คือชนิดของตัวแปรที่เก็บในคอลัมน์หนึ่ง ๆ เช่น คอลัมน์ `tip` เก็บข้อมูลชนิด `float` ขนาด 64 บิต หรือคอลัมน์ `size` เก็บข้อมูล `int` ขนาด 64 บิต เป็นต้น สำหรับใน `Pandas` มีการนิยามตัวแปรเพิ่มขึ้นอีกหลายชนิด โดยที่สำคัญที่คำสั่ง `DataFrame.info()` ก็คือ ตัวแปรชนิด `object` ซึ่งสามารถเก็บข้อมูลได้แบบผสม (mixed-type)

### จะเข้าถึงเฉพาะบางคอลัมน์ที่ต้องการได้อย่างไร

**แบบที่ 1** เข้าถึงคอลัมน์เดียวด้วยชื่อคอลัมน์

In [ ]:
s = tips['smoker']
s

0       No
1       No
2       No
3       No
4       No
      ... 
240    Yes
241    Yes
242     No
243    Yes
244     No
Name: smoker, Length: 245, dtype: object

หรือ ใช้เครื่องหมายจุด ตามด้วยชื่อคอลัมน์ก็ได้

In [ ]:
s = tips.smoker
s

0       No
1       No
2       No
3       No
4       No
      ... 
240    Yes
241    Yes
242     No
243    Yes
244     No
Name: smoker, Length: 245, dtype: object

**แบบที่ 2** เข้าถึงหลายคอลัมน์ด้วยรายการของชื่อคอลัมน์

In [ ]:
df = tips[['smoker','time','sex']]
df

smoker    time     sex
0       No  Dinner  Female
1       No  Dinner    Male
2       No  Dinner    Male
3       No  Dinner    Male
4       No  Dinner  Female
..     ...     ...     ...
240    Yes  Dinner  Female
241    Yes  Dinner    Male
242     No  Dinner    Male
243    Yes   Lunch  Female
244     No  Dinner  Female

[245 rows x 3 columns]

**คำถาม** ค่า `tips['smoker']` และ `tips[['smoker']]` เป็นตัวแปรประเภทเดียวกันหรือไม่ นักศึกษาตรวจสอบได้อย่างไร?

In [ ]:
print(type(tips['tip']))
print(type(tips[['tip']]))
print(type(tips[['tip','time','sex']]))

<class 'pandas.core.series.Series'>
<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.frame.DataFrame'>


**ข้อสังเกตุทางเทคนิค** `DataFrame` vs. `Series`

ตัวแปรประเภท `DataFrame` ไม่เหมือนกับตัวแปรชนิด `Series` กล่าวคือ `Series` คือตัวแปรที่เก็บข้อมูล 1 มิติ แต่ `DataFrame` เก็บข้อมูล 2 มิติ ดังแสดงได้จากคำสั่งต่อไปนี้

In [ ]:
s = tips['sex']
df = tips[['sex']]
print(f'shape of s = {s.shape}')
print(f'shape of df = {df.shape}')
print(f'len of s = {len(s)}')
print(f'len of df = {len(df)}')

shape of s = (245,)
shape of df = (245, 1)
len of s = 245
len of df = 245


จะเห็นว่า shape ของ `Series` เป็น `tuple` 1 มิติ แต่ของ `DataFrame` เป็น 2 มิติ ดังนั้น ในทางปฏิบัติ ควรจะทราบว่าตัวแปรที่กำลังใช้งานเป็น `Series` หรือ `DataFrame` เพื่อที่จะเข้าใจการทำงานได้ถูกต้อง

ถ้าหากต้องการแปลง `Series` ที่มีอยู่แล้ว ให้เป็น `DataFrame` (ที่มี 1 คอลัมน์) ก็สามารถทำได้โดยใช้คำสั่ง `Series.to_frame()`

In [ ]:
s.to_frame()

sex
0    Female
1      Male
2      Male
3      Male
4    Female
..      ...
240  Female
241    Male
242    Male
243  Female
244  Female

[245 rows x 1 columns]

### ต้องการ List ของชื่อคอลัมน์และชื่อ index

In [ ]:
tips.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size'], dtype='object')

In [ ]:
print(tips.columns.tolist())  # ให้ชนิดข้อมูลเป็น List ของ Python
print(list(tips.columns))

['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size']
['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size']


In [ ]:
tips.index  # จะเห็นว่า index ของแถวเป็นเลข autorun

RangeIndex(start=0, stop=245, step=1)

In [ ]:
print(tips.index.tolist())  # ให้ชนิดข้อมูลเป็น List ของ Python
print(list(tips))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221,

**ข้อสังเกตุ** `tips.columns` และ `tips.index` ไม่ใช่ตัวแปร List ดั้งเดิมในภาษา `Python` แต่เป็นชนิดตัวแปรที่นิยามไว้ใน `Pandas` เอง ดังนั้น หากต้องการแปลงเป็น List อาจใช้คำสั่ง `.tolist()` หรือไม่ก็ `list()` ก็ได้

คำสั่ง `.tolist()` ยังใช้ได้กับค่าอื่น ๆ ได้ เช่น แปลง `Series` ให้เป็น List ของ Python เป็นต้น

In [ ]:
s = tips['sex']
print(s.tolist())

['Female', 'Male', 'Male', 'Male', 'Female', 'Male', 'Male', 'Male', 'Male', 'Male', 'Male', 'Female', 'Male', 'Male', 'Female', 'Male', 'Female', 'Male', 'Female', 'Male', 'Male', 'Female', 'Female', 'Male', 'Male', 'Male', 'Male', 'Male', 'Male', 'Female', 'Male', 'Male', 'Female', 'Female', 'Male', 'Male', 'Male', 'Female', 'Male', 'Male', 'Male', 'Male', 'Male', 'Male', 'Male', 'Male', 'Male', 'Male', 'Male', 'Male', 'Male', 'Female', 'Female', 'Male', 'Male', 'Male', 'Male', 'Female', 'Male', 'Male', 'Male', 'Male', 'Male', 'Male', 'Male', 'Male', 'Female', 'Female', 'Male', 'Male', 'Male', 'Female', 'Female', 'Female', 'Female', 'Male', 'Male', 'Male', 'Male', 'Male', 'Male', 'Male', 'Female', 'Male', 'Male', 'Female', 'Male', 'Male', 'Male', 'Male', 'Male', 'Male', 'Female', 'Female', 'Female', 'Male', 'Male', 'Male', 'Male', 'Male', 'Female', 'Female', 'Female', 'Female', 'Female', 'Male', 'Male', 'Male', 'Male', 'Female', 'Male', 'Female', 'Male', 'Male', 'Female', 'Female', '

## เกี่ยวกับสถิติอย่างง่ายและการนับจำนวน

### ต้องการทราบค่าสถิติเบื้องต้น

คำถามที่ต้องการทราบในเบื้องต้น

+ พนักงานเสิร์ฟได้ทิปเฉลี่ยกี่ดอลลาร์?
+ เขาได้ทิปสูงสุดกี่ดอลลาร์?
+ กลุ่มลูกค้าที่เขาเคยเสิร์ฟมากันมากที่สุดกี่คน
+ จากชุดข้อมูลนี้ เขาทำเงินจากทิปมาแล้วกี่ดอลลาร์

In [ ]:
tips.describe()

total_bill         tip        size
count  245.000000  245.000000  245.000000
mean    19.758245    2.994204    2.567347
std      8.894722    1.382272    0.949846
min      3.070000    1.000000    1.000000
25%     13.280000    2.000000    2.000000
50%     17.780000    2.880000    2.000000
75%     24.080000    3.550000    3.000000
max     50.810000   10.000000    6.000000

ฟังก์ชัน `DataFrame.describe()` ข้างต้น แสดงผลเป็น `DataFrame` ที่สรุปค่าทางสถิติเบื้องต้น อย่างไรก็ตาม แม้ว่าข้อมูลทิปมีทั้งหมด 7 คอลัมน์ แต่ค่าสถิติที่รายงาน มีเฉพาะคอลัมน์ที่เก็บชนิดข้อมูลตัวเลข

โดยทั่วไป มักจำเป็นต้องการนำค่าทางสถิติไปใช้คำนวณต่อ, ต้องการคำนวณค่าสถิติเพียงบางค่า, หรือต้องการคำนวณค่าสถิติอื่น ๆ เช่น `sum` ที่ `DataFrame.describe()` ไม่ได้มีให้

วิธีการคำนวณทำ ดังนี้

#### คำนวณค่าสถิติของ `Series`

กรณีที่เข้าถึงเพียงคอลัมน์เดียว แต่ละคอลัมน์ก็คือตัวแปรประเภท `Series` สามารถคำนวณค่าสถิติได้ ดังนี้

In [ ]:
tips['tip'].mean()  # ค่าเฉลี่ย (ได้เป็น float)

2.994204081632653

In [ ]:
tips['tip'].std()  # ค่า SD โดย Default จะใช้สูตร SD ที่มี degree of freedom = 1 หรือก็คือค่า SD ของกลุ่มตัวอย่าง (ไม่ใช่ SD ของประชากร)

1.382272080930877

In [ ]:
tips['tip'].sum()  # ผลรวม

733.5799999999999

สำหรับค่าสถิติอื่น ๆ เช่น

- `tips['tip'].max()`
- `tips['tip'].min()`
- `tips['tip'].median()`
- `tips['tip'].count()`

เนื่องจากค่าสถิติมีหลากหลาย จึงสามารถสั่งให้คำนวณในคราวเดียวได้ ด้วยคำสั่ง `DataFrame.agg()`

In [ ]:
tips['tip'].agg(['sum','min','max','mean','std','median'])

sum       733.580000
min         1.000000
max        10.000000
mean        2.994204
std         1.382272
median      2.880000
Name: tip, dtype: float64

#### การคำนวณค่าสถิติของ `DataFrame`

In [ ]:
tips[['tip','total_bill']].mean()  # ค่าเฉลี่ยได้เป็น Series

tip            2.994204
total_bill    19.758245
dtype: float64

In [ ]:
tips[['tip','total_bill']].agg(['mean','std','max'])  # ค่าเฉลี่ยได้เป็น DataFrame

tip  total_bill
mean   2.994204   19.758245
std    1.382272    8.894722
max   10.000000   50.810000

### อยากรู้เกี่ยวกับค่าที่เป็นไปได้ในแต่ละคอลัมน์

- คอลัมน์ `time` เก็บค่าอะไรไว้บ้าง และกี่ค่า?


โดยตัวอย่างต่อไปนี้ แสดงการใช้ฟังก์ชันของ `Series`


**แบบที่ 1** ต้องการทราบแต่ค่า

In [ ]:
x = tips['time'].unique()
print(x)

['Dinner' 'Lunch']


**แบบที่ 2** ต้องการทราบแต่จำนวนแบบ

In [ ]:
x = tips['time'].nunique()
print(x)

2


**แบบที่ 3** ต้องการทราบจำนวนในแต่ละแบบ

In [ ]:
x = tips['time'].value_counts()  # ได้ผลลัพธ์เป็น Series
print(x)

Dinner    176
Lunch      69
Name: time, dtype: int64


จากการลองนับค่าในคอลัมน์ `time` พบว่ามีค่าที่เป็นไปได้ 2 แบบคือ `Dinner` และ `Lunch` ซึ่งสมเหตุสมผล เพราะนักศึกษาทราบจากพนักงานเสิร์ฟว่า เขาทำงานสองช่วงเวลา คือ ตอนเที่ยงและตอนเย็น

นอกจากนี้ คำสั่ง `value_counts()` ยังสามารถใช้ได้กับ `DataFrame` เช่นกัน ยกตัวอย่างเช่น

In [ ]:
x = tips[['sex','time']].value_counts()  # ได้ผลลัพธ์เป็น Series เช่นกัน แต่เป็น Series ที่มี Index หลายชั้น (Multi-Index)
print(x)
print(f'ข้อมูลที่รายงานจาก value_counts() เป็นชนิด {type(x)}')

sex     time  
Male    Dinner    124
Female  Dinner     52
        Lunch      36
Male    Lunch      33
dtype: int64
ข้อมูลที่รายงานจาก value_counts() เป็นชนิด <class 'pandas.core.series.Series'>


In [ ]:
x.index

MultiIndex([(  'Male', 'Dinner'),
            ('Female', 'Dinner'),
            ('Female',  'Lunch'),
            (  'Male',  'Lunch')],
           names=['sex', 'time'])

### การสร้างตารางแบบสองทาง (Two-way table)

สมมติว่าต้องการตอบคำถามต่อไปนี้

- ผู้ชายจ่ายทิปมื้อเที่ยงหรือมื้อดินเนอร์มากกว่ากัน? มีวันไหนจ่ายมากเป็นพิเศษไหม? แล้วถ้าเทียบกับผู้หญิงล่ะ?

ก็อาจสร้างตารางแบบสองทาง เพื่อนับจำแนกความถี่ของค่าแต่ละคู่ที่สนใจด้วยคำสั่ง `pd.crosstab`

**ตัวอย่างที่ 1** ไขว้ระหว่างค่าในคอลัมน์ `sex` และ `time`

In [ ]:
sex = tips['sex']
time = tips['time']
x = pd.crosstab(index=sex, columns=[time])
x

time    Dinner  Lunch
sex                  
Female      52     36
Male       124     33

In [ ]:
print(x.columns)
print(x.index)

Index(['Dinner', 'Lunch'], dtype='object', name='time')
Index(['Female', 'Male'], dtype='object', name='sex')


**ข้อสังเกตุ** index ของแถว ไม่จำเป็นต้องเป็นตัวเลขก็ได้ ดังเช่นในกรณีของตารางไขว้ข้างต้น ที่เป็น `DataFrame` ที่มี index คือ 'Female' กับ 'Male'

**ตัวอย่างที่ 2** ไขว้ระหว่างค่าในคอลัมน์ `sex` และ `(day, time)`

In [ ]:
sex = tips['sex']
day = tips['day']
time = tips['time']
x = pd.crosstab(index=sex, columns=[day, time])
x

day       Fri          Sat    Sun   Thur      
time   Dinner Lunch Dinner Dinner Dinner Lunch
sex                                           
Female      5     4     28     18      1    32
Male        7     3     59     58      0    30

In [ ]:
print(x.columns)
print(x.index)

MultiIndex([( 'Fri', 'Dinner'),
            ( 'Fri',  'Lunch'),
            ( 'Sat', 'Dinner'),
            ( 'Sun', 'Dinner'),
            ('Thur', 'Dinner'),
            ('Thur',  'Lunch')],
           names=['day', 'time'])
Index(['Female', 'Male'], dtype='object', name='sex')


## เกี่ยวกับการเลือกแถวด้วยของ `DataFrame` ด้วยตรรกะ `True/False`

ใน `Pandas` เมื่อนำข้อมูล 1 คอลัมน์ซึ่งเป็นชนิด `Series` มาเปรียบเทียบกับค่า ๆ หนึ่ง จะได้ผลลัพธ์เป็นค่าตรรกะ `True` หรือ `False` ดังตัวอย่างเช่น

In [ ]:
tips['tip'] <= 1.1

0       True
1      False
2      False
3      False
4      False
       ...  
240    False
241    False
242    False
243    False
244    False
Name: tip, Length: 245, dtype: bool

In [ ]:
tips['day'] == 'Sun'

0       True
1       True
2       True
3       True
4       True
       ...  
240    False
241    False
242    False
243    False
244    False
Name: day, Length: 245, dtype: bool

การเปรียบเทียบ `Series` กับค่า ๆ หนึ่ง (หรือจะกับอีกคอลัมน์หนึ่ง) จะได้ผลลัพธ์เป็น `Series` ที่เก็บข้อมูลชนิด `bool` (`True` หรือ `False`) โดย `Series` ดังกล่าวมีประโยชน์คือ สามารถนำไปใช้เลือกเฉพาะบางแถวในตารางได้ โดยจะกรองได้แถวที่มีค่าใน `Series` เป็น `True`

สมมติว่ามีข้อมูลใน `DataFrame` ชื่อ `df` และ `Series` ชื่อ `flags` ที่เป็นชนิด `bool` จะสามารถการกรองค่าแถวของ `df` ด้วยตรรกะ `flags` ได้ 2 แบบ คือ

1. `df.loc[flags]`
2. `df[flags]`   (เป็นรูปแบบย่อของแบบที่ 1 ซึ่งอาจสับสนกับการเข้าถึงคอลัมน์ในหัวข้อ Q1.5 ได้ แต่ก็เป็นที่นิยมใช้)


### ต้องการเฉพาะแถวที่มีค่า `day` เท่ากับ `Sun`

In [ ]:
flags = (tips['day'] == 'Sun' )  # ตัวชื่อตัวแปรว่า flags เพราะว่าถ้า True ก็หมายถึงยกธงเลือก, False ก็หมายถึงไม่เลือก
tips.loc[flags]  # แบบเต็มใช้ .loc[]

total_bill   tip     sex smoker  day    time  size
0         16.99  1.01  Female     No  Sun  Dinner     2
1         10.34  1.66    Male     No  Sun  Dinner     3
2         21.01  3.50    Male     No  Sun  Dinner     3
3         23.68  3.31    Male     No  Sun  Dinner     2
4         24.59  3.61  Female     No  Sun  Dinner     4
..          ...   ...     ...    ...  ...     ...   ...
186       20.90  3.50  Female    Yes  Sun  Dinner     3
187       30.46  2.00    Male    Yes  Sun  Dinner     5
188       18.15  3.50  Female    Yes  Sun  Dinner     3
189       23.10  4.00    Male    Yes  Sun  Dinner     3
190       15.69  1.50    Male    Yes  Sun  Dinner     2

[76 rows x 7 columns]

In [ ]:
tips[flags]   # แบบย่อ โดยละ .loc

total_bill   tip     sex smoker  day    time  size
0         16.99  1.01  Female     No  Sun  Dinner     2
1         10.34  1.66    Male     No  Sun  Dinner     3
2         21.01  3.50    Male     No  Sun  Dinner     3
3         23.68  3.31    Male     No  Sun  Dinner     2
4         24.59  3.61  Female     No  Sun  Dinner     4
..          ...   ...     ...    ...  ...     ...   ...
186       20.90  3.50  Female    Yes  Sun  Dinner     3
187       30.46  2.00    Male    Yes  Sun  Dinner     5
188       18.15  3.50  Female    Yes  Sun  Dinner     3
189       23.10  4.00    Male    Yes  Sun  Dinner     3
190       15.69  1.50    Male    Yes  Sun  Dinner     2

[76 rows x 7 columns]

**ลองทำ:** ตรวจสอบความเข้าใจ ด้วยการลองคำสั่งต่อไปนี้

| คำสั่ง | ผลลัพธ์ |
|:--|:--|
|`tips['day']` | ให้คอลัมน์ชื่อ `day` |  
|`tips[0]` | Error เพราะไม่มีคอลัมน์ชื่อ `0`  |
|`tips['Sun']` | Error เพราะไม่มีคอลัมน์ชื่อ `Sun` |
|`tips[tips['day']=='Sun']]` | ให้แถวที่มี `day` เท่ากับ `Sun` |
|`tips.loc[tips['day']=='Sun']]` |  ให้แถวที่มี `day` เท่ากับ `Sun` |   

การเข้าถึงค่าใน `DataFrame` ใน `Pandas` มีรายละเอียดมากพอสมควร ดังนี้ ณ จุดนี้ จะขอเพียงเท่านี้ก่อน


### ต้องการเฉพาะแถวที่มีค่า `tip` น้อยกว่าหรือเท่ากับ `1.1`


**วิธีที่ 1** เปรียบเทียบด้วยเครื่องหมาย `<=`

In [ ]:
flags = (tips['tip'] <= 1.1)
tips[flags]

total_bill   tip     sex smoker  day    time  size
0         16.99  1.01  Female     No  Sun  Dinner     2
67         3.07  1.00  Female    Yes  Sat  Dinner     1
92         5.75  1.00  Female    Yes  Fri  Dinner     2
111        7.25  1.00  Female     No  Sat  Dinner     1
215       12.90  1.10  Female    Yes  Sat  Dinner     2
236       12.60  1.00    Male    Yes  Sat  Dinner     2

**วิธีที่ 2** เปรียบเทียบด้วยเครื่องหมาย `>` แล้วใช้นิเสธ `~`

การกลับตรรกของ `Series` ชนิด `bool` ให้เป็นทางตรงกันข้าม ทำได้โดยใช้สัญลักษณ์ `~` (ไม่ใช่ `not`)

In [ ]:
flags = (tips['tip'] > 1.1)
tips[~flags]         # ผลลัพธ์เทียบเท่ากับการเทียบ <= 1.1

total_bill   tip     sex smoker  day    time  size
0         16.99  1.01  Female     No  Sun  Dinner     2
67         3.07  1.00  Female    Yes  Sat  Dinner     1
92         5.75  1.00  Female    Yes  Fri  Dinner     2
111        7.25  1.00  Female     No  Sat  Dinner     1
215       12.90  1.10  Female    Yes  Sat  Dinner     2
236       12.60  1.00    Male    Yes  Sat  Dinner     2

### ต้องการเฉพาะแถวที่มี `tip` ในช่วง 1.25 ถึง 1.35 ด้วยตรรกะ `&`


การกระทำระหว่าง `Series` ชนิด `bool` 2 `Series` จะใช้สัญลักษณ์ `&` (ไม่ใช่ `and`)

In [ ]:
flags_1 = (tips['tip'] >= 1.25)
flags_2 = (tips['tip'] <= 1.35)
tips[flags_1 & flags_2]

total_bill   tip     sex smoker   day    time  size
43         9.68  1.32    Male     No   Sun  Dinner     2
75        10.51  1.25    Male     No   Sat  Dinner     2
135        8.51  1.25  Female     No  Thur   Lunch     2
235       10.07  1.25    Male     No   Sat  Dinner     2

### ต้องการเฉพาะแถวที่มี `tip` ที่น้อยกว่าเท่ากับ 1.1 หรือมากกว่าเท่ากับ 9.0 ด้วยตรรกะ `|`

การกระทำระหว่าง `Series` ชนิด `bool` 2 `Series` จะใช้สัญลักษณ์ `|` (ไม่ใช่ `or`)


In [ ]:
flags_1 = (tips['tip'] <= 1.1)
flags_2 = (tips['tip'] >= 9.0)
tips[flags_1 | flags_2]

total_bill    tip     sex smoker  day    time  size
0         16.99   1.01  Female     No  Sun  Dinner     2
67         3.07   1.00  Female    Yes  Sat  Dinner     1
92         5.75   1.00  Female    Yes  Fri  Dinner     2
111        7.25   1.00  Female     No  Sat  Dinner     1
170       50.81  10.00    Male    Yes  Sat  Dinner     3
212       48.33   9.00    Male     No  Sat  Dinner     4
215       12.90   1.10  Female    Yes  Sat  Dinner     2
236       12.60   1.00    Male    Yes  Sat  Dinner     2

ทั้งนี้ ตรรกกะ  `(tip <= 1.1) & (tip >= 9.0)` เทียบเท่ากับ `~((tip > 1.1) | (tip < 9.0))`

In [ ]:
flags_1 = (tips['tip'] > 1.1)
flags_2 = (tips['tip'] < 9.0)
tips[~(flags_1 & flags_2)]

total_bill    tip     sex smoker  day    time  size
0         16.99   1.01  Female     No  Sun  Dinner     2
67         3.07   1.00  Female    Yes  Sat  Dinner     1
92         5.75   1.00  Female    Yes  Fri  Dinner     2
111        7.25   1.00  Female     No  Sat  Dinner     1
170       50.81  10.00    Male    Yes  Sat  Dinner     3
212       48.33   9.00    Male     No  Sat  Dinner     4
215       12.90   1.10  Female    Yes  Sat  Dinner     2
236       12.60   1.00    Male    Yes  Sat  Dinner     2

โดยสรุป โอเปอร์เรเตอร์ในการเปรียบเทียบ ที่เป็นไปได้มีดังนี้

| เปรียบเทียบ | ความหมาย |
|:--:|:-- |
| `==` | เท่ากันหรือไม่ |
| `!=` | ไม่เท่ากันหรือไม่ |
| `>` หรือ `>=` | มากกว่า, มากกว่าเท่ากับ |
| `<` หรือ `<=` | น้อยกว่า, น้อยกว่าเท่ากับ |

| ตรรกะ | ความหมาย |
|:--:|:-- |
| `~` | นิเสธ |
| `&` | และ |
| `\|` | หรือ |

### **เลือกแถวด้วยตรรกะ แต่อยากได้เพียงบางคอลัมน์**

จาก Q3.1 ถึง Q3.4 นักศึกษาสามารถเลือกแถวตามเงื่อนไขที่ต้องการได้แล้ว แต่ไม่อยากได้ทุกคอลัมน์ ก็สามารถทำได้ 2 แบบ ดังตาราง

| แบบที่  | คำสั่ง | ยกตัวอย่าง | คำแนะนำ |
|:--:|:--:|:-- |:-- |
|1 | `df.loc[flags, ชื่อคอลัมน์หรือ list ชื่อคอลัมน์]`|  `tips.loc[flags, 'sex']`  | ควรใช้ |
|2 | `df.loc[flags][ชื่อคอลัมน์หรือ list ชื่อคอลัมน์]`|  `tips.loc[flags]['sex']` | ไม่ควรใช้ |

โดยที่ `df` คือ `DataFrame` และมีและ `flags` เป็น `Series` ชนิด `bool`

**ระวัง:** ผู้หัดใช้ `Pandas` มือใหม่ มักใช้แบบที่ 2 ซึ่งเป็นแบบควรหลีกเลี่ยง แม้ว่าจะได้แถวที่ต้องการเหมือนกัน ซึ่งความแตกต่างระหว่างสองแบบนี้ จะเกิดขึ้นในกรณีที่ต้องการเขียนค่าลงใน `DataFrame` (ซึ่งเกินเนื้อหาของคาบนี้) ดังนั้นจึงขอให้นึกศึกษาทำความคุ้นเคยกับแบบที่ 1 ไว้ก่อน แต่หากอยากใช้แบบที่ 2 ก็ย่อมได้ แต่ขอให้ทราบถึงข้อจำกัดของมัน





In [ ]:
flags = (tips['tip'] > 1.1)
tips.loc[flags]['sex']  # ไม่ควรใช้แบบนี้

1        Male
2        Male
3        Male
4      Female
5        Male
        ...  
240    Female
241      Male
242      Male
243    Female
244    Female
Name: sex, Length: 239, dtype: object

In [ ]:
flags = (tips['tip'] > 1.1)
tips.loc[flags, 'sex']  # ควรใช้แบบนี้

1        Male
2        Male
3        Male
4      Female
5        Male
        ...  
240    Female
241      Male
242      Male
243    Female
244    Female
Name: sex, Length: 239, dtype: object

หรือกรณีที่ต้องการมากกว่า 1 คอลัมน์

In [ ]:
flags = (tips['tip'] > 1.1)
tips.loc[flags, ['total_bill','sex']]  # ต้องการ 2 คอลัมน์

total_bill     sex
1         10.34    Male
2         21.01    Male
3         23.68    Male
4         24.59  Female
5         25.29    Male
..          ...     ...
240       27.18  Female
241       22.67    Male
242       17.82    Male
243       13.00  Female
244       18.78  Female

[239 rows x 2 columns]

ทั้งนี้ คำสั่ง
```python
df.loc[flags, ชื่อคอลัมน์หรือ list ชื่อคอลัมน์]
```
ไม่สามารถละ `.loc` ได้ ไม่เหมือนกับที่ `df[flags]` เป็นรูปย่อของ `df.iloc[flags]` ดังนั้น

```python
tips[flags, 'sex']      # คำสั่งนี้ Error (ขอให้นักศึกษาลองทำดู)
```
จะไม่สามารถทำงานได้

### **มีแถวที่ซ้ำกันบ้างไหมนะ**

สามารถดูแถวที่ซ้ำด้วย `DataFrame.duplicated()` โดยจะให้ผลลัพธ์เป็น `Series` ชนิด `bool` โดยแถวที่ซ้ำทั้งแถวกับแถวอื่น ก็จะมีค่าเป็น `True` หากแถวนั้นไม่ซ้ำกับแถวไหนเลย ก็จะมีค่าเป็น `False` (เป็นกรณีที่กำหนดพารามิเตอร์ `keep=False` หากกำหนดให้ `keep` เป็นค่าอื่น ก็จะแตกต่างออกไป)

In [ ]:
flags = tips.duplicated(keep=False)
tips[flags]

total_bill  tip     sex smoker   day   time  size
198        13.0  2.0  Female    Yes  Thur  Lunch     2
202        13.0  2.0  Female    Yes  Thur  Lunch     2
243        13.0  2.0  Female    Yes  Thur  Lunch     2

*จงตอบคำถาม:*
1. พารามิเตอร์ `keep` มีค่าที่เป็นไปได้ 3 แบบ คือ `False`, `'first'`, และ `'last'` แล้วแต่ละแบบแตกต่างกันอย่างไร
2. หากไม่ระบุ `keep` แล้ว `Pandas` จะได้ผลอย่างไร

`DataFrame.duplicated()` จะพิจารณาว่าทั้งแถวซ้ำกันหรือไม่ แต่ในกรณีที่ต้องการให้พิจารณาเฉพาะเพียงบางคอลัมน์ในแถว ก็สามารถระบุได้ด้วยพารามิเตอร์ `subset`

เช่น (อาจด้วยเหตุผลบางประการ) สมมติว่าสนใจคู่ `day` และ `time` แล้วต้องการดึงเฉพาะแถวที่ปรากฎคู่นี้ครั้งแรก
 ก็สามารถทำได้ ดังนี้

In [ ]:
flags = tips.duplicated(subset=['day','time'])  # แถวที่มีคู่ day และ time ซ้ำกันจะถูก flag ว่า True เว้นแต่เป็นคู่แรกที่เจอ
flags = ~flags # กลับตรรกะเพื่อเลือกเฉพาะแถวแรกที่ซ้ำ
tips[flags]

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
19        20.65  3.35    Male     No   Sat  Dinner     3
77        27.20  4.00    Male     No  Thur   Lunch     4
90        28.97  3.00    Male    Yes   Fri  Dinner     2
220       12.16  2.20    Male    Yes   Fri   Lunch     2
244       18.78  3.00  Female     No  Thur  Dinner     2

จากตารางข้างต้น ตีความได้ว่า คู่ `(Thur,	Lunch)` ถูกพบครั้งแรกที่แถว `index=77` เป็นต้น

### ตรรกะ OR และ AND ของทั้ง `Series` ด้วย `any()` และ `all()`

หากต้องการเอาค่าทั้งหมดของ `Series` มากระทำการ OR กันทุกค่า เช่น `Series` ชื่อ `a` ที่เป็นชนิด `bool` ที่มีสมาชิก `a1`,`a2`,`a3`,`a4` อยู่ 4 ค่า (len=4, ซึ่งแต่ละค่าเป็นแค่ `True/False`) และต้องการหาค่า `a1 OR a2 OR a3 OR a4` จะสามารถทำได้โดยใช้คำสั่ง `Series.any()`

เช่น ต้องการตรวจว่าในคอลัมน์ชื่อ `'day'` มีสักหนึ่งแถวที่มีค่า `Thu` หรือไม่ จะเขียนคำสั่ง ดังนี้

In [ ]:
(tips['day'] == 'Thur').any()

True

ซึ่งจะได้จริง เพราะพบอย่างน้อยหนึ่งแถว

หรือในกรณีที่ต้องการหาค่า `a1 AND a2 AND a3 AND a4` ก็ทำได้โดยใช้คำสั่ง `Series.all()`

เช่น ต้องการตรวจว่า ค่าทิปในตาราง มีทิปที่ต่ำกว่า 1 USD หรือไม่ ก็สามารถตรวจโดยเขียนคำสั่ง ดังนี้

In [ ]:
(tips['tip'] < 1).all()

False

ซึ่งแสดงว่า ในตารางค่าทิปมีค่าไม่น้อยกว่า 1 USD ทุกค่า

## **เกี่ยวกับการสร้างคอลัมน์ใหม่**

### ต้องการแปลงค่าเงินจากดอลลาร์เป็นบาท

ข้อมูลทิปที่พนักงานเสิร์ฟให้มีค่าทิปและค่าบิลอาหารเป็นหน่วย USD แต่เนื่องจากนักศึกษาอาจอยากทราบเป็นค่าเงินบาท ก็สามารถนำค่าจากคอลัมน์เก่าไปคำนวณ แล้วเก็บในคอลัมน์ใหม่ได้

In [ ]:
exchange_rate = 35.0 # สมมติอัตราแลกเปลี่ยนเป็น 1 USD เท่ากับ 35 THB
tips['tip']*exchange_rate   # ค่าอัตราแลกเปลี่ยนถูกคูณเข้าทุก ๆ ค่าของ tip

0       35.35
1       58.10
2      122.50
3      115.85
4      126.35
        ...  
240     70.00
241     70.00
242     61.25
243     70.00
244    105.00
Name: tip, Length: 245, dtype: float64

In [ ]:
tips['tip_thb'] = tips['tip']*exchange_rate  # เก็บผลการคำนวณไว้ในคอลัมน์ใหม่ที่ชื่อ tip_thb
tips.head()

total_bill   tip     sex smoker  day    time  size  tip_thb
0       16.99  1.01  Female     No  Sun  Dinner     2    35.35
1       10.34  1.66    Male     No  Sun  Dinner     3    58.10
2       21.01  3.50    Male     No  Sun  Dinner     3   122.50
3       23.68  3.31    Male     No  Sun  Dinner     2   115.85
4       24.59  3.61  Female     No  Sun  Dinner     4   126.35

ทั้งนี้ อาจเลี่ยงการสร้างคอลัมน์ใหม่ แต่เก็บผลการคำนวณทับในคอลัมน์เดิมก็ได้ ดังเช่น

`tips['tip'] = tips['tip']*exchange_rate`

แต่ไม่แนะนำให้ทำลักษณะนี้ใน Google Colab เนื่องจากอาจเกิดการอัพเดตทบค่าเทอม หากมีการสั่งให้โค้ดในบรรทัดนี้ทำงานหลายครั้ง

**ลองทำ:** จงสร้างคอลัมน์ชื่อ `total_bill_thb` ที่เป็นการแปลงค่าเงินจากคอลัมน์ `total_bill` ให้เป็นหน่วยบาท

### ทิปที่ให้เป็นกี่เปอร์เซนต์ของค่าอาหาร

ที่สหรัฐอเมริกา การให้ค่าทิปจะไม่ถูกรวมอยู่ในบิลอาหาร แต่เป็นการให้ด้วยความสมัครใจ ดังนั้นจึงอยากทราบว่าแต่ละครั้ง ลูกค้าจ่ายบิลกี่เปอร์เซนต์ของค่าอาหาร ซึ่งการคำนวณทำได้ดังนี้

In [ ]:
tips['tip']/tips['total_bill']    # การนำ Series มาหาร Series จะได้ผลเป็นการนำค่าที่เลข index ตรงกันมาหารกัน

0      0.059447
1      0.160542
2      0.166587
3      0.139780
4      0.146808
         ...   
240    0.073584
241    0.088222
242    0.098204
243    0.153846
244    0.159744
Length: 245, dtype: float64

In [ ]:
tips['%tip'] = tips['tip']/tips['total_bill']*100      # คูณ 100 เพื่อแปลงเป็น %
tips.head()

total_bill   tip     sex smoker  day    time  size  tip_thb       %tip
0       16.99  1.01  Female     No  Sun  Dinner     2    35.35   5.944673
1       10.34  1.66    Male     No  Sun  Dinner     3    58.10  16.054159
2       21.01  3.50    Male     No  Sun  Dinner     3   122.50  16.658734
3       23.68  3.31    Male     No  Sun  Dinner     2   115.85  13.978041
4       24.59  3.61  Female     No  Sun  Dinner     4   126.35  14.680765

### ทิปแต่ละวันเป็นกี่เปอร์เซนต์ของทิปทั้งหมด

สมมติว่าพนักงานคนนี้ลาออกจากร้านอาหารนี้แล้ว และเขาอยากทราบว่า ค่าทิปในแต่ละวัน คิดเป็นกี่เปอร์เซนต์ของค่าทิปทั้งหมดที่เขาเคยได้รับมา

In [ ]:
tips['tip']/tips['tip'].sum()

0      0.001377
1      0.002263
2      0.004771
3      0.004512
4      0.004921
         ...   
240    0.002726
241    0.002726
242    0.002386
243    0.002726
244    0.004090
Name: tip, Length: 245, dtype: float64

In [ ]:
tips['%tip_of_tips'] = tips['tip']/tips['tip'].sum()*100
tips.head()

total_bill   tip     sex smoker  day    time  size  tip_thb       %tip  \
0       16.99  1.01  Female     No  Sun  Dinner     2    35.35   5.944673   
1       10.34  1.66    Male     No  Sun  Dinner     3    58.10  16.054159   
2       21.01  3.50    Male     No  Sun  Dinner     3   122.50  16.658734   
3       23.68  3.31    Male     No  Sun  Dinner     2   115.85  13.978041   
4       24.59  3.61  Female     No  Sun  Dinner     4   126.35  14.680765   

   %tip_of_tips  
0      0.137681  
1      0.226288  
2      0.477112  
3      0.451212  
4      0.492107

**ลองทำ:** จงตรวจสอบว่าคอลัมน์ `%tip_of_tips` มีผลรวมเท่ากับ 1 จริง ๆ หรือไม่


## **เกี่ยวกับการเรียงลำดับข้อมูล**

การเรียงลำดับจะใช้คำสั่ง `DataFrame.sort_values()` โดยมีพารามิเตอร์ที่สำคัญคือ

- `by` เพื่อระบุว่าให้เรียงโดยใช้ค่าในคอลัมน์ใด
- `ascending` เพื่อระบุว่าให้เรียงจากน้อยไปมากหรือไม่ (default: `True`)

ตัวอย่างของการเรียงข้อมูล เป็นดังนี้




### เรียงลำดับข้อมูล `tip` จากน้อยไปมาก

In [ ]:
sorted_tips = tips.sort_values(by='tip', ascending=True)   # ต้องการเรียงลำดับแถวตามค่า tip จากน้อยไปมาก
sorted_tips.head(8)

total_bill   tip     sex smoker  day    time  size  tip_thb       %tip  \
111        7.25  1.00  Female     No  Sat  Dinner     1    35.00  13.793103   
92         5.75  1.00  Female    Yes  Fri  Dinner     2    35.00  17.391304   
236       12.60  1.00    Male    Yes  Sat  Dinner     2    35.00   7.936508   
67         3.07  1.00  Female    Yes  Sat  Dinner     1    35.00  32.573290   
0         16.99  1.01  Female     No  Sun  Dinner     2    35.35   5.944673   
215       12.90  1.10  Female    Yes  Sat  Dinner     2    38.50   8.527132   
237       32.83  1.17    Male    Yes  Sat  Dinner     2    40.95   3.563814   
235       10.07  1.25    Male     No  Sat  Dinner     2    43.75  12.413108   

     %tip_of_tips  
111      0.136318  
92       0.136318  
236      0.136318  
67       0.136318  
0        0.137681  
215      0.149950  
237      0.159492  
235      0.170397

โดยผลจากการเรียงลำดับ จะเห็นว่าค่า `index` ใน `sorted_tips` มีการสลับลำดับกัน

### เรียงลำดับข้อมูล `tip` และ `total_bill` (ตามลำดับ) จากน้อยไปมาก

In [ ]:
sorted_tips = tips.sort_values(by=['tip','total_bill'], ascending=True )   # จะเห็นว่าเรียงค่าตาม tip โดยที่ค่า tip เท่ากัน จะเรียงค่า total_bill
sorted_tips

total_bill    tip     sex smoker   day    time  size  tip_thb       %tip  \
67         3.07   1.00  Female    Yes   Sat  Dinner     1    35.00  32.573290   
92         5.75   1.00  Female    Yes   Fri  Dinner     2    35.00  17.391304   
111        7.25   1.00  Female     No   Sat  Dinner     1    35.00  13.793103   
236       12.60   1.00    Male    Yes   Sat  Dinner     2    35.00   7.936508   
0         16.99   1.01  Female     No   Sun  Dinner     2    35.35   5.944673   
..          ...    ...     ...    ...   ...     ...   ...      ...        ...   
141       34.30   6.70    Male     No  Thur   Lunch     6   234.50  19.533528   
59        48.27   6.73    Male     No   Sat  Dinner     4   235.55  13.942407   
23        39.42   7.58    Male     No   Sat  Dinner     4   265.30  19.228818   
212       48.33   9.00    Male     No   Sat  Dinner     4   315.00  18.621974   
170       50.81  10.00    Male    Yes   Sat  Dinner     3   350.00  19.681165   

     %tip_of_tips  
67       0.136318  
92       0.136318  
111      0.136318  
236      0.136318  
0        0.137681  
..            ...  
141      0.913329  
59       0.917419  
23       1.033289  
212      1.226860  
170      1.363178  

[245 rows x 10 columns]

### อยากได้ `index` ที่เรียงลำดับตามค่า `index`

จากตัวแปร `sorted_tips` ที่ค่า `index` สลับที่กันนั้น หากต้องการเรียงลำดับค่า `index` ก็สามารถทำได้ด้วยคำสั่ง `DataFrame.sort_index()`

In [ ]:
sorted_tips.sort_index()

total_bill   tip     sex smoker   day    time  size  tip_thb       %tip  \
0         16.99  1.01  Female     No   Sun  Dinner     2    35.35   5.944673   
1         10.34  1.66    Male     No   Sun  Dinner     3    58.10  16.054159   
2         21.01  3.50    Male     No   Sun  Dinner     3   122.50  16.658734   
3         23.68  3.31    Male     No   Sun  Dinner     2   115.85  13.978041   
4         24.59  3.61  Female     No   Sun  Dinner     4   126.35  14.680765   
..          ...   ...     ...    ...   ...     ...   ...      ...        ...   
240       27.18  2.00  Female    Yes   Sat  Dinner     2    70.00   7.358352   
241       22.67  2.00    Male    Yes   Sat  Dinner     2    70.00   8.822232   
242       17.82  1.75    Male     No   Sat  Dinner     2    61.25   9.820426   
243       13.00  2.00  Female    Yes  Thur   Lunch     2    70.00  15.384615   
244       18.78  3.00  Female     No  Thur  Dinner     2   105.00  15.974441   

     %tip_of_tips  
0        0.137681  
1        0.226288  
2        0.477112  
3        0.451212  
4        0.492107  
..            ...  
240      0.272636  
241      0.272636  
242      0.238556  
243      0.272636  
244      0.408953  

[245 rows x 10 columns]

### อยากเอา `index` ไปสร้างเป็นคอลัมน์ใหม่

มีความเป็นไปได้ที่จะนำค่า `index` นั้นไปสร้างเป็นคอลัมน์ใหม่ ด้วยคำสั่ง `DataFrame.reset_index()`

In [ ]:
sorted_tips.reset_index()    # จะเห็นว่ามีคอลัมน์ชื่อ index เพิ่มเข้ามา

index  total_bill    tip     sex smoker   day    time  size  tip_thb  \
0       67        3.07   1.00  Female    Yes   Sat  Dinner     1    35.00   
1       92        5.75   1.00  Female    Yes   Fri  Dinner     2    35.00   
2      111        7.25   1.00  Female     No   Sat  Dinner     1    35.00   
3      236       12.60   1.00    Male    Yes   Sat  Dinner     2    35.00   
4        0       16.99   1.01  Female     No   Sun  Dinner     2    35.35   
..     ...         ...    ...     ...    ...   ...     ...   ...      ...   
240    141       34.30   6.70    Male     No  Thur   Lunch     6   234.50   
241     59       48.27   6.73    Male     No   Sat  Dinner     4   235.55   
242     23       39.42   7.58    Male     No   Sat  Dinner     4   265.30   
243    212       48.33   9.00    Male     No   Sat  Dinner     4   315.00   
244    170       50.81  10.00    Male    Yes   Sat  Dinner     3   350.00   

          %tip  %tip_of_tips  
0    32.573290      0.136318  
1    17.391304      0.136318  
2    13.793103      0.136318  
3     7.936508      0.136318  
4     5.944673      0.137681  
..         ...           ...  
240  19.533528      0.913329  
241  13.942407      0.917419  
242  19.228818      1.033289  
243  18.621974      1.226860  
244  19.681165      1.363178  

[245 rows x 11 columns]

แต่อย่างไรก็ตาม ถ้าหากไม่ต้องการนำ `index` ไปสร้างเป็นคอลัมน์ใหม่ แต่ให้ลบทิ้งเลย ก็สามาถทำได้ด้วยการเพิ่มพารามิเตอร์ `drop=True`

In [ ]:
sorted_tips.reset_index(drop=True)     # ลบลำดับก่อนการเรียงข้อมูลออกไป

total_bill    tip     sex smoker   day    time  size  tip_thb       %tip  \
0          3.07   1.00  Female    Yes   Sat  Dinner     1    35.00  32.573290   
1          5.75   1.00  Female    Yes   Fri  Dinner     2    35.00  17.391304   
2          7.25   1.00  Female     No   Sat  Dinner     1    35.00  13.793103   
3         12.60   1.00    Male    Yes   Sat  Dinner     2    35.00   7.936508   
4         16.99   1.01  Female     No   Sun  Dinner     2    35.35   5.944673   
..          ...    ...     ...    ...   ...     ...   ...      ...        ...   
240       34.30   6.70    Male     No  Thur   Lunch     6   234.50  19.533528   
241       48.27   6.73    Male     No   Sat  Dinner     4   235.55  13.942407   
242       39.42   7.58    Male     No   Sat  Dinner     4   265.30  19.228818   
243       48.33   9.00    Male     No   Sat  Dinner     4   315.00  18.621974   
244       50.81  10.00    Male    Yes   Sat  Dinner     3   350.00  19.681165   

     %tip_of_tips  
0        0.136318  
1        0.136318  
2        0.136318  
3        0.136318  
4        0.137681  
..            ...  
240      0.913329  
241      0.917419  
242      1.033289  
243      1.226860  
244      1.363178  

[245 rows x 10 columns]

### ต้องการบันทึก `DataFrame` ลงไฟล์ `csv`

หากต้องการบันทึกตัวแปร `DataFrame` เก็บลงไฟล์ ดังเช่นตัวแปร `sorted_tips` ที่เรียงลำดับค่าไว้แล้ว ก็สามารถทำได้ง่ายด้วย `DataFrame.to_csv()`

ทั้งนี้ กำหนดพารามิเตอร์ `index=False` เพื่อที่จะไม่ต้องบันทึกเลข `index` ลงในไฟล์ `csv` (อย่าลืมว่าเลขนี้ auto generate มาจากตอนอ่านไฟล์ `csv` เข้ามา)

In [ ]:
# ไฟล์จะถูกเซฟไว้ใน Google Colab (เข้าถึงจากเมนูรูปโฟลเดอร์ด้านซ้ายมือ)
sorted_tips.to_csv('lnwzaa55+.csv', index=False)

## เกร็ดในการใช้ `Pandas` (Method chaining)

เพื่อความกระชับในการเรียกใช้ฟังก์ชันของ `Pandas` อาจเลือกที่จะเรียกใช้ฟังก์ชันต่อกันเป็นทอด ๆ (Method chaining) โดยไม่ต้องสร้างตัวแปรใหม่ก็ได้ ดังตัวอย่างต่อไปนี้

In [ ]:
tips[['total_bill','tip']].sort_values(by='tip', ascending=False).sample(50).agg(['max','min','mean']).to_csv('ggez.csv')

**คำถาม:** โค้ดข้างต้นทำอะไร?

การเรียกใช้ฟังก์ชันเป็นทอด ๆ ข้างต้น ค่อนข้างที่จะอ่านยาก ดังนั้น ในทางปฏิบัติ อาจใช้วงเล็บมาครอบคำสั่ง เพื่อทำให้สามารถเขียนโค้ดได้ในหลายบรรทัด ดังนี้

In [ ]:
# ข้อสังเกตคือ มีวงเล็บมาครอบ
(tips[['total_bill','tip']]
   .sort_values(by='tip', ascending=False)
   .sample(50, random_state=888)
   .agg(['max','min','mean'])
   .to_csv('ggez.csv')
)

## การวาดกราฟเบื้องต้น

`Pandas` ไม่สามารถวาดกราฟได้ด้วยตัวเอง แต่ช่วยอำนวยความสะดวกการวาดกราฟข้อมูลได้ โดยอาศัยความสามารถของ [`matplotlib`](https://matplotlib.org/), [`Plotly Express`](https://plotly.com/python/plotly-express/) หรือไลบรารีอื่น ๆ ที่ทำงานอยู่เบื้องหลัง (backend)


### การใช้ `Plotly` โดยผ่าน `Pandas`

โดย default แล้ว หากไม่ตั้งค่าอะไร `Pandas` ก็จะใช้ `matplotlib` อย่างไรก็ตาม `plotly` สามารถวาดกราฟได้สวยงามกว่า และมีลูกเล่นมากกว่า ซึ่งหากต้องการใช้ `plotly` เป็นเบื้องหลัง (backend) ในการวาดกราฟ จะต้องกำหนดค่าดังนี้


In [ ]:
pd.options.plotting.backend = "plotly"

การใช้ฟังก์ชันวาดกราฟของ `Pandas` ที่ใช้ backend เป็น `plotly` นั้น เมื่อใช้คำสั่ง `plot()` [(คู่มือ)](https://plotly.com/python/pandas-backend/) ดังนี้

```
DataFrame.plot(x=..., y=..., kind=..., title=..., labels=..., width=..., height=...)
```
ก็จะสามารถวาดกราฟได้หลายชนิด สำหรับรายละเอียดเบื้องต้นของค่าพารามิเตอร์ของฟังก์ชัน  แสดงดังตารางต่อไปนี้

|พารามิเตอร์ | ค่า | ความหมาย|
|:---|:---|:---|
| `x=` | `str` | ชื่อคอลัมน์ของข้อมูลแกนนอน |
| `y=` | `str` | ชื่อคอลัมน์ของข้อมูลพิกัดแกนตั้ง  |
| `color=` | `str` | ชื่อคอลัมน์ที่ใช้จัดกลุ่มสี |
| `kind=`| `'line'` | line plot (default)|
|| `'bar'` | bar plot|
|| `'hist'` | histogram|
|| `'scatter'` | scatter plot |
|`title=` | `str` | ชื่อกราฟ|
|`labels=`| `dict()` | กำหนดชื่อแกน |
| `width=` | `int` | ความกว้างของกราฟ |
| `height=` | `int` | ความสูงของกราฟ |

ค่าอื่น ๆ ที่เป็นไปได้ สำหรับ `kind=` ของ `plotly` backend คือ `'box'`,`'area'`,`'violin'`, `'strip'`, `'funnel'`, `'density_heatmap'`, `'density_contour'`

ทั้งนี้ กรณีเลือก `matplotlib` เป็น backend ก็อาจมีค่าพารามิเตอร์บางตัว แตกต่างจากนี้

#### Bar chart

ตัวอย่างต่อไปนี้เป็นการวาดกราฟแท่งที่มีแกนนอนคือ `day` และความสูงคือ `total_bill`

In [ ]:
tips.plot(x="day", y="total_bill", kind='bar')

สามารถเพิ่มเติมค่า `color` เพื่อให้แยกสีตามค่า `sex`

In [ ]:
tips.plot(x="day", y="total_bill", color="sex", kind='bar')

การกำหนด `color='sex'` จะได้กราฟที่แต่ละแท่งซ้อนกัน เรียกว่า stacked bar chart ซึ่งสามารถแยกแต่ละแท่งให้เป็นแท่งเดียว ด้วยการกำหนด `barmode='group'` ดังนี้

In [ ]:
tips.plot(x="day", y="total_bill", color="sex", kind='bar', barmode='group')

**ลองทำ**
- ลองเพิ่มพารามิเตอร์ `orientation='h'` พร้อมสลับค่าของ `x=` และ `y=` จะได้กราฟอะไร
- ลองเปลี่ยนเป็น `kind='hist'` จะได้กราฟอะไร? แล้ว `kind='bar'` และ `kind='hist'` ต่างกันอย่างไร?


#### Histogram

In [ ]:
tips.plot(x="%tip", kind='hist', nbins=40, text_auto=True)

**ลองทำ** ลองปรับค่า `nbins`

#### Scatter plot

In [ ]:
fig = tips.plot(x='total_bill', y='%tip',
          color='time',   # กำหนดสีแต่ละจุดตามค่า time
          kind='scatter', # ชนิดกราฟ
          title='กราฟแสดงทิปที่ได้ในแต่ละครั้ง',
          labels={'%tip': 'สัดส่วนการให้ทิป (%)','total_bill':'ค่าบิลอาหาร (USD)'},  # กำหนด dictionary ในการตั้งชื่อแกนนอนและแกนตั้ง
          width=800, height=400,  # ขนาดของรูป
)
fig.update_layout(margin_t=40)  # แก้ไขระยะระหว่าง title กับกราฟ
fig.show()

### การใช้ `Plotly` โดยตรง

`plotly` backend สำหรับ `Pandas` เป็นคุณสมบัติใหม่ที่เพิ่งมีใน `Pandas` และยังไม่สมบูรณ์ ดังนั้น การใช้งานจึงมีข้อจำกัด ดังนั้น ในหลายกรณีจึงมีความจำเป็นที่จะต้องวาดกราฟโดยเรียกใช้  `plotly` โดยตรง ซึ่งทำได้โดยการนำเข้า `plotly` ด้วยคำสั่ง ดังนี้

```
import plotly.express as px
```

และแทนที่จะใช้คำสั่ง `plot()` ในการวาดกราฟแต่ละชนิด แต่จะใช้คำสั่งต่อไปนี้

| ฟังก์ชัน | ชนิดกราฟ |
|:---|:---|
| `px.bar(data, x=..., y=...)` | กราฟแท่ง (หาก `orientation='h'`) จะได้กราฟแท่งแนวนอน |
| `px.scatter(data, x=..., y=...)` | scatter plot  |
| `px.histogram(data, x=...)` | histogram |

ซึ่งยกตัวอย่างการสร้างกราฟแท่ง ได้แก่

In [ ]:
import plotly.express as px
px.bar(tips, x="day", y="total_bill")

จะเห็นว่ากราฟที่ได้ เหมือนกับการวาดกราฟโดยเรียกใช้ฟังก์ชัน `plot()` จาก `Pandas`

**ลองทำ**

- กราฟที่เคยวาดโดยใช้ฟังก์ชัน `plot()` ของ `Pandas` จงวาดกราฟแบบเดิม แต่เปลี่ยนไปใช้คำสั่ง `px.bar()`, `px.scatter()` และ `px.histogram()`

### ศึกษาตัวอย่างการวาดกราฟด้วยตนเอง

เนื่องจากการวาดกราฟมีรูปแบบมากมาย นักศึกษาอาจศึกษาการวาดกราฟจากการดูตัวอย่างและคู่มือของ `plotly` ซึ่งขอให้ไว้ตามรายการต่อไปนี้

1. [การใช้ `plotly` เป็น backend ของ `Pandas`](https://plotly.com/python/pandas-backend/)
2. [ตัวอย่างการใช้ `plotly`](https://plotly.com/python/plotly-express/)
  - [ตัวอย่างการสร้าง bar chart](https://plotly.com/python/bar-charts/) และ [คู่มือ](https://plotly.com/python-api-reference/generated/plotly.express.bar)
  - [ตัวอย่างการสร้าง histogram](https://plotly.com/python/histograms/) และ [คู่มือ](https://plotly.com/python-api-reference/generated/plotly.express.histogram)
  - [ตัวอย่างการสร้าง scatter plot](https://plotly.com/python/line-and-scatter/) และ [คู่มือ](https://plotly.com/python-api-reference/generated/plotly.express.scatter)


*หมายเหตุ:* อาจดู[คู่มือการใช้ฟังก์ชัน `plot()` ของ `Pandas`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.plot.html) แต่บางพารามิเตอร์อาจใช้ไม่ได้ เนื่องจากเป็นคู่มือสำหรับ backend ที่เป็น `matplotlib`

**จบเนื้อหา**

นักศึกษาลองสรุปซิว่า วันนี้นักศึกษาได้เรียนรู้คำสั่งอะไรบ้าง